// @doptools/tslib-cli-core/dist/lib/lib/commands/SchematicsCommandBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchematicsCommandBase = void 0;
const SchematicsSimpleFs_1 = require("../files/simple-filesystem/SchematicsSimpleFs");
const runner_1 = require("../schematics/runner");
const CommandBase_1 = require("./CommandBase");
class SchematicsCommandBase extends CommandBase_1.CommandBase {
    constructor() {
        super(...arguments);
        this.fs = new SchematicsSimpleFs_1.SchematicsSimpleFs();
        this.readonly = true;
    }
    buildOptions() {
        return {
            ...this.arguments,
            ...this.flags,
            mode: this.readonly ? 'ro' : 'rw'
        };
    }
    async runRules(rules, opts = {}) {
        return runner_1.RuleRunner.create(opts).execute(rules);
    }
    async runRule(rule, opts = {}) {
        return this.runRules([rule], opts);
    }
    run() {
        const options = this.buildOptions();
        const r = this.runRule((tree, context) => {
            this.fs.tree = tree;
            return this.execute(tree, context, options);
        }, options);
        this.fs.tree = null;
        return r;
    }
}
exports.SchematicsCommandBase = SchematicsCommandBase;
