// @doptools/cli/lib/commands/plugins/list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const tslib_cli_core_1 = require("@doptools/tslib-cli-core");
const cli_ux_1 = tslib_1.__importDefault(require("cli-ux"));
const plugins_1 = require("../../util/plugins");
function getPluginType(cliPkg, plugin) {
    var _a;
    const pluginDef = (_a = cliPkg.oclif.plugins) === null || _a === void 0 ? void 0 : _a.find(p => typeof p === 'string' ? p === plugin.name : p.name === plugin.name);
    if (!pluginDef) {
        return plugin.name === '@doptools/cli' ? 'main' : 'unknown';
    }
    if (typeof pluginDef === 'string') {
        return 'core';
    }
    return pluginDef.type;
}
let PluginsListCommand = class PluginsListCommand extends tslib_cli_core_1.CommandBase {
    constructor() {
        super(...arguments);
        this.all = false;
    }
    async run() {
        const cliPkg = await plugins_1.readCliPackageJson();
        let plugins = this.config.plugins
            .map(plugin => ({
            name: plugin.name,
            type: getPluginType(cliPkg, plugin),
            valid: plugin.valid,
            version: plugin.version
        }));
        if (!this.all) {
            plugins = plugins.filter(_ => _.type !== 'main' && _.type !== 'core');
        }
        cli_ux_1.default.table(plugins, {
            name: {},
            type: { minWidth: 8 },
            version: {}
        });
    }
};
tslib_1.__decorate([
    tslib_cli_core_1.BooleanFlag(),
    tslib_1.__metadata("design:type", Boolean)
], PluginsListCommand.prototype, "all", void 0);
PluginsListCommand = tslib_1.__decorate([
    tslib_cli_core_1.CliCommand({})
], PluginsListCommand);
exports.default = PluginsListCommand;
